src/asn1.ml

```diff
@@ -615,11 +667,263 @@ struct
+    ; 0x7F49, (`PUBLIC_KEY, true)
+    ; 0x5F20, (`HOLDER_REFERENCE, false)
+    ; 0x7F4C, (`HOLDER_AUTH_TEMPLATE, true)
+    ; 0x5F25, (`EFFECTIVE_DATE, false)
+    ; 0x5F24, (`EXPIRATION_DATE, false)
+    ; 0x5F37, (`SIGNATURE, false)
+    ; 0x06, (`OID, false)
+    ; 0x42, (`CA_REFERENCE, false)
+    ; 0x67, (`REQ_AUTHENTICATION, true)
+    ; 0x53, (`ROLE_AND_ACCESS_RIGHTS, false)
+    ; 0x81, (`MODULUS, false)
+    ; 0x82, (`EXPONENT, false)
+    ; 0x82, (`COEFFICIENT_A, false)
+    ; 0x83, (`COEFFICIENT_B, false)
+    ; 0x84, (`BASE_POINT_G, false)
+    ; 0x85, (`BASE_POINT_R_ORDER, false)
+    ; 0x86, (`PUBLIC_POINT_Y, false)
+    ; 0x87, (`COFACTOR_F, false)
+    ]
+
+  let find_cvc_object_type tag =
+    let code =
+      Cstruct.get_char tag 0
+      |> Char.code
+    in
+    try code, List.assoc code cvc_object_types
+    with Not_found ->
+      let code =
+        let msb = code * 0x100 in
+        let lsb =
+          Cstruct.get_char tag 1
+          |> Char.code
+        in
+        msb + lsb
+      in
+      code, List.assoc code cvc_object_types
+
+  (* utility function to parse a big-endian blob as a Z.t *)
+  let atoz_bigendian s =
+    let reverse s =
+      let n = String.length s in
+      String.init n (fun i -> s.[n-1-i])
+    in
+    Z.of_bits @@ reverse @@ Cstruct.to_string s
+
+  let grammar =
+    let open Asn in
+    let f = function
+      | oid when oid = rsa_oid -> RSA
+      | oid when oid = ecdsa_oid -> ECDSA
+      | oid -> Unknown oid in
+    let g = function
+      | RSA -> rsa_oid
+      | ECDSA -> ecdsa_oid
+      | Unknown oid -> oid in
+    map f g oid
+
+  let decode_oid str =
+    let t, left = Asn.(decode_exn (codec ber grammar) str) in
+    t
+
+  let decode bytes =
+    let buffer = Cstruct.create 4_096 in
+    (* FSM to produce `Type ..., `Length ..., `Value ... tokens from a blob.
+     * This tries to exploit tailcall recursion as much as possible in order to
+     * avoid a stack explosion
+     *)
+    let rec tokenize ~acc bytes i lim state = function
+      | Init ->
+          if i >= lim then List.rev acc
+          else tokenize ~acc bytes i lim None Type
+      | Type ->
+          Cstruct.blit bytes i buffer 0 2;
+          let cvc_type = find_cvc_object_type buffer in
+          let acc = `Type cvc_type :: acc in
+          begin match cvc_type with
+            | tag, _ when tag <= 0xff ->
+                let i = i + 1 in
+                (tokenize[@tailcall]) ~acc bytes i lim (Some cvc_type) Length
+            | tag, _ ->
+                let i = i + 2 in
+                (tokenize[@tailcall]) ~acc bytes i lim (Some cvc_type) Length
+          end
+      | Length ->
+          let code = Char.code (Cstruct.get_char bytes i) in
+          if code < 0x80
+          then begin
+            let i = i + 1 in
+            (tokenize[@tailcall]) ~acc:((`Length code) :: acc) bytes i lim state (Value code)
+          end
+          else
+            begin match code with
```

Comment:
Do you mean a function such as this?

```ocaml
parse_len ?(i = 0) cs =
  let code = Cstruct.get_uint8 cs i in
  if code < 0x80
  then begin
    (* ... *)
  end
```

I'm not sure if that would end up making things much clearer, as I don't think it's used anywhere else, and leaving it inline with the parser allows the eventual extension to arbitrarily long length specifiers.